### PR TITLE
[SPARK-14334] [SQL] add toLocalIterator for Dataset/DataFrame

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

RDD.toLocalIterator() could be used to fetch one partition at a time to reduce the memory usage. Right now, for Dataset/Dataframe we have to use df.rdd.toLocalIterator, which is super slow also requires lots of memory (because of the Java serializer or even Kyro serializer).

This PR introduce an optimized toLocalIterator for Dataset/DataFrame, which is much faster and requires much less memory. For a partition with 5 millions rows, `df.rdd.toIterator` took about 100 seconds, but df.toIterator took less than 7 seconds. For 10 millions row, rdd.toIterator will crash (not enough memory) with 4G heap, but df.toLocalIterator could finished in 12 seconds.

The JDBC server has been updated to use DataFrame.toIterator.
## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by davies** [comment]
cc @marmbrus @rxin

**Comment by rxin** [comment]
hm we need to think about the api -- this is a scala iterator right? it's going to be problematic for java users.

maybe we can have toLocalScalaIterator and toLocalJavaIterator, or just have toLocalIterator returning a Java iterator, and scala users can easily do the implicit conversion or using asScala anyway.

**Comment by davies** [comment]
@rxin Good point, will use the Java iterator as public API.

**Comment by rxin** [comment]
cc @sameeragarwal for review

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 1]
```diff
@@ -71,6 +71,7 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
     assert(ds.first() == item)
     assert(ds.take(1).head == item)
     assert(ds.takeAsList(1).get(0) == item)
+    assert(ds.toLocalIterator().next === item)
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: nit: next()

**Comment by sameeragarwal** [comment]
LGTM

**Pull request closed without merging**